mock-tikv/src/store.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use kvproto::kvrpcpb::KvPair;
+use std::{
+    collections::HashMap,
+    sync::{Arc, RwLock},
+};
+
+#[derive(Debug, Clone)]
+pub struct KvStore {
+    data: Arc<RwLock<HashMap<Vec<u8>, Vec<u8>>>>,
+}
+
+impl Default for KvStore {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl KvStore {
+    pub fn new() -> KvStore {
+        KvStore {
+            data: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    pub fn raw_get(&self, key: &[u8]) -> Vec<u8> {
+        let data = self.data.read().unwrap();
+        data.get(key).unwrap_or(&vec![]).to_vec()
+    }
+
+    pub fn raw_batch_get(&self, keys: &[Vec<u8>]) -> Vec<KvPair> {
+        let data = self.data.read().unwrap();
+        keys.iter()
+            .map(|key| {
+                let mut pair = KvPair::default();
+                pair.set_value(data.get(key).unwrap_or(&vec![]).to_vec());
+                pair.set_key(key.to_vec());
+                pair
+            })
+            .collect::<Vec<_>>()
+    }
+
+    pub fn raw_put(&self, key: &[u8], value: &[u8]) {
+        let mut data = self.data.write().unwrap();
+        *data.entry(key.to_vec()).or_default() = value.to_vec();
+    }
+
+    pub fn raw_batch_put(&self, pairs: &[KvPair]) {
+        let mut data = self.data.write().unwrap();
+        for pair in pairs {
+            *data.entry(pair.get_key().to_vec()).or_default() = pair.get_value().to_vec();
+        }
+    }
+
+    // if success, return the key deleted
+    pub fn raw_delete(&self, key: &[u8]) -> Result<Vec<u8>, ()> {
+        let mut data = self.data.write().unwrap();
+        data.remove(key).ok_or(())
+    }
+
+    // if any of the key does not exist, return non-existent keys
+    pub fn raw_batch_delete<'a>(&self, keys: &'a [Vec<u8>]) -> Result<(), Vec<&'a str>> {
+        let mut data = self.data.write().unwrap();
+        let non_exist_keys = keys
+            .iter()
+            .filter(|&key| !data.contains_key(key))
+            .map(|key| std::str::from_utf8(key).unwrap())
+            .collect::<Vec<_>>();
+        if !non_exist_keys.is_empty() {
+            Err(non_exist_keys)
+        } else {
+            keys.iter().for_each(|key| {
```

Comment:
Can you rewrite the function so that you only iterate over `keys` once?

Reply:
I don't come up with an elegant way to do this.

Reply:
It's a bit of a novelty question really. My attempt is:

```
        let mut data = self.data.write().unwrap();
        let non_exist: Vec<_> = keys.into_iter().filter_map(|k| data.remove(&k).xor(Some(k))).collect();
        if non_exist.is_empty() {
            Ok(())
        } else {
            Err(non_exist)
        }
```

Note that it has slightly different semantics - it removes any keys which are in the batch, whereas yours doesn't delete anything if any key is missing. I'm not actually sure what TiKV does, but I suspect it will not matter in practice (given how MySQL `DELETE` works).

Reply:
Yeah if we don't want to gurantee the atomicity of this operation, we can do it.